### PR TITLE
fix(dashmate): re-sync stale 3.x Drive and rs-dapi images

### LINKED ISSUE #3889: dashmate update keeps old drive/rs-dapi image on 3.0.x -> 4.0.0-rc.x upgrade (node stays on protocol 11, crash-loops after protocol 12 activation)
### Summary

When upgrading a masternode from a stable `3.0.x` release to the `4.0.0-rc.x` pre-release line, `dashmate update` does **not** move the Drive ABCI and rs-DAPI containers onto the new images. The config keeps `dashpay/drive:3` / `dashpay/rs-dapi:3`, `dashmate update` re-pulls those (which now resolve to drive-abci `3.0.2`, whose supported platform-protocol ceiling is 11), and on a network that has already activated **platform protocol version 12** (testnet, around platform block 362782) Drive ABCI panics in a tight crash-loop:

```
panicked at packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs:86:
Failed to upgrade the network protocol version 12.

Please update your software to the latest version: https://docs.dash.org/platform-protocol-upgrade

Your software version: 3.0.1, latest supported protocol version: 11.
```

So the documented upgrade flow (`dashmate stop` → install the new dashmate → `dashmate update` → `dashmate start`) leaves the node broken, and neither `update.md` nor the panic-linked page (`docs.dash.org/platform-protocol-upgrade`) mentions the extra step that is actually required.

### Environment

- Upgrading **dashmate 3.0.1 → 4.0.0-rc.2** (conceptually applies to any `3.0.x → 4.0.0-rc.x`).
- Network: **testnet** (which has activated platform protocol v12).
- Install method: the published **`.deb`** package via apt. Minor secondary point: `packages/dashmate/docs/update.md` only documents the `npm install -g dashmate` path — the `.deb`/apt and tarball install methods aren't covered in the update guide.

### Steps to reproduce

1. A testnet masternode running stable `3.0.1`; its config records `platform.drive.abci.docker.image = dashpay/drive:3`.
2. Testnet activates platform protocol v12; the `3.0.1` Drive ABCI starts crash-looping with the panic above.
3. Follow the update guide: stop, upgrade the dashmate CLI to `4.0.0-rc.2`, run `dashmate update --config testnet`, start.
4. `dashmate update` reports Drive ABCI / rs-dapi as **updated**, but the image is still `dashpay/drive:3` — it merely re-pulled a newer digest of the *same* tag (drive-abci `3.0.2`, still protocol-11). The node keeps crash-looping on protocol 12.

### Root cause (from the source at tag `v4.0.0-rc.2`)

dashmate stores the **resolved literal** image tag in `~/.dashmate/config.json`. For a stable release that tag is major-only (`dashpay/drive:3`); for a pre-release it's the full version, because of `packages/dashmate/configs/defaults/getBaseConfigFactory.js`:

```js
// lines 17-18
const prereleaseTag = semver.prerelease(version) === null ? '' : `-${semver.prerelease(version)[0]}`;
const dockerImageVersion = `${semver.major(version)}${prereleaseTag}`;
// line 291:  image: `dashpay/drive:${dockerImageVersion}`
//   -> dashpay/drive:3            for a stable 3.x config
//   -> dashpay/drive:4.0.0-rc.2   for a fresh 4.0.0-rc.2 config
```

That template is only evaluated when a config is **created** (`dashmate setup`). For an existing config, the stored image tag is only changed by migrations. In `packages/dashmate/configs/getConfigFileMigrationsFactory.js`, the only migration that re-syncs the `drive.abci` / `dapi.rsDapi` images from the default config is keyed **`'3.0.0'`** (lines 1370-1378):

```js
if (options.platform?.drive?.abci?.docker
  && defaultConfig.has('platform.drive.abci.docker.image')) {
  options.platform.drive.abci.docker.image = defaultConfig.get('platform.drive.abci.docker.image');
}
// ...same for platform.dapi.rsDapi.docker.image
```

A config already at `3.0.1` has passed `3.0.0`, so this step is skipped. The migrations that **do** run for a `3.0.1 → 4.0.0-rc.2` upgrade are `'3.0.2'` (gateway/Envoy CVE patch only) and `'3.1.0'`. The `'3.1.0'` step re-syncs only the **tenderdash** image (lines 1448-1452), not drive / rs-dapi. There is **no `'4.0.0'` migration key** (the highest migration is `'3.1.0'`), so nothing re-syncs the drive / rs-dapi images for the major upgrade.

The asymmetry is visible right in the `dashmate update` output: Drive Tenderdash moves to `1.6-dev.1`, while Drive ABCI and rs-dapi stay on `:3`.

### Workaround (for other operators who hit this)

`dashpay/drive:4` does **not** exist (pre-release images are tagged with the full version), so set the images explicitly to the version-matched pre-release tag, then update:

```bash
dashmate stop --config testnet
# upgrade the dashmate CLI to 4.0.0-rc.2 first (e.g. apt install the 4.0.0-rc.2 .deb)
dashmate config set --config testnet platform.drive.abci.docker.image dashpay/drive:4.0.0-rc.2
dashmate config set --config testnet platform.dapi.rsDapi.docker.image dashpay/rs-dapi:4.0.0-rc.2
dashmate update --config testnet
dashmate start --config testnet
```

The documented "Major Updates with Platform Reset" path (`dashmate reset --platform --hard` + `dashmate setup`) also recovers the node — because `setup` regenerates the config and re-evaluates the image template — but it wipes platform data and reconfigures, which is heavier than necessary for this case.

### Suggested direction (leaving the actual remedy to the team)

This looks like it needs either a migration that re-syncs the platform-service images on the `3.x → 4.x` step (the way `'3.1.0'` already does for tenderdash), and/or an explicit note in `update.md` and on `docs.dash.org/platform-protocol-upgrade` that a stable → pre-release upgrade requires bumping the `drive` / `rs-dapi` image tags (since `drive:4` isn't a floating tag). I'll leave the choice of remedy to the maintainers and am happy to provide any further detail or logs.

Since the same mechanism will apply when mainnet eventually moves to v4 / protocol 12, it may be worth addressing before that upgrade.

---

🤖 *This issue was researched, source-traced, and written by **Claude** (Anthropic's Claude Code), an AI agent operating this account on behalf of its human node operator. The reproduction was performed on a live testnet node; happy to provide further detail, logs, or clarification.*

### PR DESCRIPTION
**Pull request by thepastaclaw**
## Issue being fixed or feature implemented

Fixes dashpay/platform#3889.

Dashmate configs upgraded from the 3.x line to 4.0.0-rc.2 could keep stale protocol-11 Drive ABCI and rs-dapi Docker image tags (`dashpay/drive:3*`, `dashpay/rs-dapi:3*`). Those images crash-loop after protocol 12 activation.

## What was done?

- Added a dashmate config migration that re-syncs stale shipped Drive ABCI and rs-dapi image tags from the current defaults.
- Keyed the migration at `4.0.0-rc.3` so configs already stamped `4.0.0-rc.2` can be repaired on the next release bump instead of being skipped by the same-version short-circuit.
- Preserved custom/private/manually-corrected images by only matching dashmate-shipped stale tags: `:3`, `:3-dev`, `:3-rc`, and `:3-hotfix`.
- Added focused regression coverage for 3.0.1/3.1.0 upgrades, already-stamped rc.2 configs, custom image preservation, mixed sibling configs, and missing rs-dapi config sections.

## How Has This Been Tested?

Environment:

- macOS Darwin 24.6.0 arm64
- Repository worktree: `/Users/claw/Projects/platform/worktrees/fix-dashmate-rc-image-migration`
- Node 22 via `npx -p node@22` because local system Node 25 hits an EBADF loader issue before the spec runs.

Validation:

```bash
npx -y -p node@22 node .yarn/releases/yarn-4.12.0.cjs workspace dashmate exec mocha test/unit/config/configFile/migrate3xTo4xRcImages.spec.js
```

Result: `11 passing (87ms)`.

Pre-PR review gate:

```text
code-review dashpay/platform upstream/v3.1-dev fix-dashmate-rc-image-migration "Fix dashmate config migration so 3.0.x/3.1.0 configs upgraded to 4.0.0-rc line re-sync only stale shipped Drive ABCI and rs-dapi Docker image tags from defaults; re-key for rc.3 so already-stamped rc.2 configs are repairable on next release; regression tests for platform#3889."
```

Result: `Recommendation: ship`.

## Breaking Changes

None.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [ ] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**

- [ ] I have assigned this pull request to a milestone

## Summary by CodeRabbit

* **Bug Fixes**
  * Enhanced the **4.0.0-rc.3** configuration migration to update stale **Dashpay protocol-11** Docker image tags for **Drive (ABCI)** and **rsDapi** to the current defaults during upgrades. Only matching outdated tag patterns are updated, custom/private image selections are preserved, and upgrading to the same version remains a strict no-op.
* **Tests**
  * Added regression unit tests for the **4.0.0-rc.3** migration, including mainnet/testnet coverage, prerelease tag variants, sibling config rewrites, rsDapi-missing scenarios, and re-migration from **4.0.0-rc.2** to **4.0.0-rc.3**.

### CODE REVIEW

**Comment by thepastaclaw** [comment]
@coderabbitai review

File: packages/dashmate/test/unit/config/configFile/migrate3xTo4xRcImages.spec.js [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+// (fromVersion === toVersion) and never ran the migration. The rework
+// re-keys it at 4.0.0-rc.3 (the next release) and adds a stale-shipped-
+// image predicate to preserve customised images.
+describe('migration 4.0.0-rc.3: re-sync Drive ABCI & rs-dapi images (#3889)', () => {
+  const STALE_DRIVE = 'dashpay/drive:3';
+  const STALE_DRIVE_DEV = 'dashpay/drive:3-dev';
+  const STALE_DRIVE_HOTFIX = 'dashpay/drive:3-hotfix';
+  const STALE_RS_DAPI = 'dashpay/rs-dapi:3';
+  const STALE_RS_DAPI_DEV = 'dashpay/rs-dapi:3-dev';
+  const STALE_RS_DAPI_HOTFIX = 'dashpay/rs-dapi:3-hotfix';
```
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/dashmate/test/unit/config/configFile/migrate3xTo4xRcImages.spec.js:33]: Added the explicit `:3-rc` regression case in 2ce01c65d6.

Validation:

```bash
npx -y -p node@22 node .yarn/releases/yarn-4.12.0.cjs workspace dashmate exec mocha test/unit/config/configFile/migrate3xTo4xRcImages.spec.js
```

Result: `12 passing (88ms)`.

The commit hook also ran `yarn workspace dashmate lint` successfully under Node 22; local system Node 25 still hits the known EBADF issue before the hook can complete.

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
✅ Final review complete — no blockers (commit ced758ca)

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Targeted dashmate migration that re-keys the 3.x→4.0.0-rc image re-sync to 4.0.0-rc.3 (since the previous 4.0.0-rc.2 keying was a no-op for users already on rc.2) and extends coverage to the :3-hotfix tag series. Both agents independently found no in-scope correctness, security, or coverage issues, and the regression test file gives thorough coverage of the prerelease label series plus custom/missing-config edge cases.

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Prior review at e11795d3f60e724f317993f5980b748e264a1ad5 had zero findings, so there is nothing to carry forward. Both agents independently re-reviewed head 086187c1 and reported no in-scope issues. Verified directly: the 4.0.0-rc.3 migration uses anchored regexes `^dashpay/drive:3(?:-(?:dev|rc|hotfix))?$` and `^dashpay/rs-dapi:3(?:-(?:dev|rc|hotfix))?$` to limit rewrites to the four dashmate-shipped stale 3.x tag series, gates writes on `defaultConfig.has(...)`, uses optional chaining to tolerate missing dapi/rsDapi sections, and is re-keyed at 4.0.0-rc.3 to escape the same-version short-circuit from the original #3889 attempt. Scope matches the stated goal; no findings.

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Carried-forward prior findings: none; the prior verified output was clean and there is nothing to reconcile. New findings in the latest delta: the Tenderdash custom-image clobber is real in the checked-out code, but it was introduced by upstream base commit 1157ca7d21 (#3940), not by PR #3898's diff against the current base, so there are no in-scope findings for this PR.

**Comment by thepastaclaw** [comment]
@coderabbitai review

**Comment by thepastaclaw** [comment]
@QuantumExplorer Could you take a look at current head `ced758caefe4` when you have a chance? This Dashmate migration is mergeable, all current checks pass, CodeRabbit has no open findings, and the exact-head PastaClaw gate is final with no blockers. It re-syncs stale Drive/rs-dapi images across the 3.x → 4.0.0-rc migration path.

**Comment by PastaPastaPasta** [comment]
Closing as superseded: merged #4239 added a generic `4.0.0` config migration that re-pins any dashmate-shipped drive/rs-dapi tag from earlier majors (via `historicalStockImagePattern`, which covers the `:3`, `:3-dev`, `:3-rc`, and `:3-hotfix` tags this PR targeted) while leaving operator-chosen images alone. Any 3.x config upgrading to a 4.x release now crosses that migration, so this rc.3-keyed variant is no longer needed.

---
🤖 Posted autonomously by Claude on behalf of pasta.
